### ISSUE TITLE
[Feature] Override/Set Variable Image Resolution (vision_soft_tokens_per_image) for Gemma 4

### ISSUE THREAD

**Issue by dwohlfahrt** [jundot/omlx#1023]
Unless I've missed it, there's currently no way to set/override `vision_soft_tokens_per_image` for gemma 4.  The default is 280, which is far too low for most real-world image tasks (OCR, data extraction, etc.).   The official [Gemma 4 docs](https://huggingface.co/google/gemma-4-26B-A4B-it#5-variable-image-resolution) state the values as 70, 140, 280, 560, and 1120, but the model will actually accept much higher values, which I depend on to get maximum OCR accuracy on old handwritten document scans.  Point being, if you are kind enough to implement this I'd very much appreciate a free text input vs limiting to the 5 suggested presets :)

FYI, support for this is [already there in mlx-vlm](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/gemma4/config.py#L139), so just need a way to configure the value and pass it through via the admin/chat UI and/or cli.  Admittedly, I haven't extensively dug through the code to see if there's a way to pass this in via a cli flag, so apologies if that's already there and I missed it.  Appreciate it and thanks for this phenomenal app!

**Comment by miabchdave** [comment]
What if you use the oMLX Admin > Model Settings > Gemma 4 > [SidePanel] Chat Template Kwargs > Add > Custom > 

Key: max_soft_tokens 
Value: 1120

Does that work for your OCR?

Edit:  Note, you must use the Google Gemma 4 accepted values as the mlx-vlm kwargs code checks for this.  The values are the following:  

_SUPPORTED_SOFT_TOKENS = (70, 140, 280, 560, 1120)

Edit:  **_Never mind_**, even though the max image tokens are supported per request, looks like the chat template doesn't support it or pass it to the processor.   So still 280 is used.

**Comment by miabchdave** [comment]
Found a workaround:

In the model folder > processor_config.json > change max_soft_tokens to one of the accepted values then reload the model.  Confirmed image prompt is now 560.  

For example:

...

    "image_processor_type": "Gemma4ImageProcessor",
    "image_seq_length": 280,
    "image_std": [
      1.0,
      1.0,
      1.0
    ],
    "max_soft_tokens": 560,
    "patch_size": 16,
    "pooling_kernel_size": 3,
    "resample": 3,
    "rescale_factor": 0.00392156862745098,
    "size": {
      "height": 224,
      "width": 224
    }

...

**Comment by dwohlfahrt** [comment]
Man, no luck :( I was actually trying the same approach earlier and every time I bump up max_soft_tokens generation errors out with this in the log:

```
2026-04-30 21:11:56,978 - omlx.server - ERROR - [-] - Error during chat streaming: [broadcast_shapes] Shapes (1,9963,1152) and (1,2520,1152) cannot be broadcast.
```

### Full environment details in case it's helpful:
model: mlx-community/gemma-4-26b-a4b-it-8bit
machine: m4 max 128GB
oMLX version: 0.3.8 (installed via the sequoia dmg)
os version: Sequoia 15.7.3

contents of `~/.omlx/models/gemma-4-26b-a4b-it-8bit/processor_config.json` (max_soft_tokens updated to 1120):
```
{
  "audio_ms_per_token": 40,
  "audio_seq_length": 750,
  "feature_extractor": {
    "dither": 0.0,
    "feature_extractor_type": "Gemma4AudioFeatureExtractor",
    "feature_size": 128,
    "fft_length": 512,
    "fft_overdrive": false,
    "frame_length": 320,
    "hop_length": 160,
    "input_scale_factor": 1.0,
    "max_frequency": 8000.0,
    "mel_floor": 0.001,
    "min_frequency": 0.0,
    "padding_side": "right",
    "padding_value": 0.0,
    "per_bin_mean": null,
    "per_bin_stddev": null,
    "preemphasis": 0.0,
    "preemphasis_htk_flavor": true,
    "return_attention_mask": true,
    "sampling_rate": 16000
  },
  "image_processor": {
    "do_convert_rgb": true,
    "do_normalize": false,
    "do_rescale": true,
    "do_resize": true,
    "image_mean": [
      0.0,
      0.0,
      0.0
    ],
    "image_processor_type": "Gemma4ImageProcessor",
    "image_seq_length": 280,
    "image_std": [
      1.0,
      1.0,
      1.0
    ],
    "max_soft_tokens": 1120,
    "patch_size": 16,
    "pooling_kernel_size": 3,
    "resample": 3,
    "rescale_factor": 0.00392156862745098
  },
  "image_seq_length": 280,
  "processor_class": "Gemma4Processor",
  "video_processor": {
    "do_convert_rgb": true,
    "do_normalize": true,
    "do_rescale": true,
    "do_resize": true,
    "do_sample_frames": true,
    "image_mean": [
      0.0,
      0.0,
      0.0
    ],
    "image_std": [
      1.0,
      1.0,
      1.0
    ],
    "max_soft_tokens": 280,
    "num_frames": 32,
    "patch_size": 16,
    "pooling_kernel_size": 3,
    "resample": 3,
    "rescale_factor": 0.00392156862745098,
    "return_metadata": false,
    "video_processor_type": "Gemma4VideoProcessor"
  }
}
```

contents of `~/.omlx/model_settings.json` (see `chat_template_kwargs` and `forced_ct_kwargs`):
```
{
  "version": 1,
  "models": {
    "gemma-4-26b-a4b-it-8bit": {
      "temperature": 1.0,
      "top_p": 0.95,
      "top_k": 64,
      "force_sampling": false,
      "chat_template_kwargs": {
        "max_soft_tokens": 1120
      },
      "forced_ct_kwargs": [
        "max_soft_tokens"
      ],
      "enable_thinking": false,
      "thinking_budget_enabled": false,
      "turboquant_kv_enabled": false,
      "turboquant_kv_bits": 4.0,
      "turboquant_skip_last": true,
      "specprefill_enabled": false,
      "dflash_enabled": false,
      "is_pinned": false,
      "is_default": true,
      "trust_remote_code": false
    }
  }
}
```

Nonetheless, appreciate you looking into this @miabchdave !

**Comment by miabchdave** [comment]
I think this error may be caused by something in 0.3.8 and Gemma4 with images.  For example, just booted up and ran this test with Unsolth's MLX 8Bit Gemma4 31B and 1120 entered in processor_config.json (the model has manually updated Google chat templates/tokenizers).  

The test worked immediately, BUT I DO get that same error when running Gemma4 31B BF16 from mlx community.  It seems maybe it's a cache related bug?  Not sure.  Also, I have the oMLX custom kwargs setting still there, not sure if related at all (see image).

<img width="940" height="629" alt="Image" src="https://github.com/user-attachments/assets/a27e45f2-af32-4eb7-9c31-37011c8b67eb" />

<img width="1228" height="93" alt="Image" src="https://github.com/user-attachments/assets/7d3e26b2-a309-4913-b10f-d6d70bf51c88" />

**Comment by miabchdave** [comment]
Note, I just converted my own BF16 Gemma 4 31B into MLX with mlx-vlm (from https://huggingface.co/google/gemma-4-31B-it) - then it works fine with **_just changing_** the max-soft-tokens in processor_config.json to 1120.  No error.  

I made sure to start a new chat in oMLX to test (works from terminal through curling the API as well).

Maybe there's something weird with the vision towers in the mlx-community Gemma 4 models?  My model safetensor files are definitely a different SHA and a few bytes different each.  

I used a newer mlx-vlm (commit e41cd25 version 0.4.5) to do the Gemma4 conversion compared to mlx-community.

**Comment by gentoothebest** [comment]
Try this: set both `default_output_length` (in config.json) and `max_soft_tokens` (in processor_config.json) to 1120.

`vision_soft_tokens_per_image` seems to do nothing.
